### PR TITLE
Move track-artist-updating code into testable function

### PR DESCRIPTION
**Pull request by derat**
Move the release editor code that updates track credits when a release's credits are changed into a new standalone getUpdatedTrackArtists function, and add unit tests for it.

No functional changes are intended, but this should make it easier to make future improvements as discussed in MBS-13273.

# Problem

I'm planning to work on MBS-13273 some more, but testing the many edge cases via Selenium is painful.

# Solution

This change moves code for updating track credits out of the release editor's `init.js` file and into a separate Flow-checked utility file that can be unit-tested.

# Testing

I added a unit test that exercises a bunch of different cases. The existing release editor Selenium tests will hopefully pass, too.

### CODE REVIEW

File: root/static/scripts/common/immutable-entities.js [hunk 1]
```diff
@@ -42,7 +42,7 @@ export const reduceArtistCredit =
  */
 export function reduceArtistCreditNames(
   names: $ReadOnlyArray<ArtistCreditNameT>,
-  dropFinalJoinPhrase?: false,
+  dropFinalJoinPhrase?: boolean = false,
```
Comment (derat) [inline root/static/scripts/common/immutable-entities.js:45]: Whoops, I think this was a bug from one of my earlier changes. Still getting used to Flow...

File: root/static/scripts/edit/utility/getUpdatedTrackArtists.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * If the track's credits should not be changed, the original trackArtists
+ * array is returned.
+ */
+export default function getUpdatedTrackArtists(
+  trackArtists: $ReadOnlyArray<ArtistCreditNameT>,
+  oldArtists: $ReadOnlyArray<ArtistCreditNameT>,
+  newArtists: $ReadOnlyArray<ArtistCreditNameT>,
+): $ReadOnlyArray<ArtistCreditNameT> {
+  const oldArtistsReduced = reduceArtistCreditNames(oldArtists);
```
Comment (derat) [inline root/static/scripts/edit/utility/getUpdatedTrackArtists.js:35]: This code previously only reduced these credits once per release, but now it does it once per track. I initially added support for passing in already-reduced credits, but it felt like premature optimization since we still need to reduce the track credits in the next statement.

File: root/static/scripts/edit/utility/getUpdatedTrackArtists.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    ) {
+      /*
+       * Replace the old release artist(s) with the new one(s) and restore
+       * the old join phrase following the release artist.
+       */
+      const names = cloneArrayDeep(newArtists)
+        .concat(trackArtists.slice(oldArtists.length));
+      names[newArtists.length - 1] = {
+        artist: names[newArtists.length - 1].artist,
+        joinPhrase: trackArtists[oldArtists.length - 1].joinPhrase,
```
Comment (derat) [inline root/static/scripts/edit/utility/getUpdatedTrackArtists.js:84]: The old code was updated the `joinPhrase` property in the existing immutable object, but I think it was just able to get away with that by virtue of being in a non-Flow-checked file.

File: root/static/scripts/tests/edit/utility/getUpdatedTrackArtists.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        name('B', '2', ' feat. '),
+        name('C', '3'),
+      ],
+      oldRel: [name('A', '1', ' & ', 'a'), name('B', '2')],
+      newRel: [name('A', '1', ' & '), name('B', '2')],
+      want: [
+        /*
+         * TODO: As discussed in MBS-13273, the first artist's credited-as
+         * name should be removed here.
+         */
```
Comment (derat) [inline root/static/scripts/tests/edit/utility/getUpdatedTrackArtists.js:119]: This is something that I plan to change, but I figured it'd be good to add a test case now to document the existing behavior.

**Review by derat — COMMENTED** [review]
@mwiencek and/or @reosarevok, mind taking a look since you reviewed the code that this is moving?

**Comment by derat** [comment]
@brainzbot, retest this please

I think the failure was this, in the "Artist Edit Form" Selenium test, which seems unrelated to this change:

```
    ⨯ caught exception: SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON at JSON.parse (<anonymous>) at handleCommand
```

File: root/static/scripts/tests/edit/utility/getUpdatedTrackArtists.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  const testCases = [
+    {
+      desc: 'artist change, track matches old',
+      track: [name('Old', '1')],
+      oldRel: [name('Old', '1')],
+      newRel: [name('New', '2')],
+      want: [name('New', '2')],
+    },
+    {
+      desc: 'artist change, track doesn\'t match old',
```
Comment (reosarevok) [inline root/static/scripts/tests/edit/utility/getUpdatedTrackArtists.js:36]: You can use unicode apostrophes rather than escaping the ASCII ones here.

Comment (mwiencek) [reply root/static/scripts/tests/edit/utility/getUpdatedTrackArtists.js:36]: The main benefit of that is if the test fails, you can grep for the string more easily, since what is printed will match the source code.

Comment (derat) [reply root/static/scripts/tests/edit/utility/getUpdatedTrackArtists.js]: Makes sense; thanks for the explanation. I've updated this.

**Review by mwiencek — APPROVED** [review]
Thanks for all of the added tests! This will definitely make further changes easier to be sure about.

**Comment by derat** [comment]
@reosarevok, any other comments? If not, mind merging this? :-)
